app/models/favorite.rb

```diff
@@ -27,9 +27,11 @@ def self.add!(post:, user:)
         raise Error, "You have already favorited this post"
       end
 
-      Favorite.create!(:user_id => user.id, :post_id => post.id)
-      Post.where(:id => post.id).update_all("fav_count = fav_count + 1")
-      post.append_user_to_fav_string(user.id)
+      post.with_lock do
```

Comment:
This entire method is pretty much broken.
- `favorites` should have a unique index on `post_id, user_id` (you can remove the separate `post_id` index) when you add this
- The transaction needs to have `serializable` isolation (`Post.transaction(isolation: :serializable)`)
- No other locking is necessary